### ISSUE TITLE
AssertConfigurationIsValid() equivilent 

### ISSUE THREAD

**Issue by tig2810** [expressmapper/ExpressMapper#24]
Hello,

Do you have or plan to implement an AssertConfigurationIsValid() equivilent, found in automapper?  This basically checks all maps are valid during build. 

Woudl be a great feature.

Thanks

**Comment by IanYates** [comment]
The handy way to use this, as I think @tig2810 is alluding to (and you probably already know), is to have a unit test where you call your various mapper setup functions from your real project and then have the mapper utility assert that all of the mapper configuration is valid.

Just curious, is the wontfix label indicating that this won't be done or just that it's a feature rather than a bug?

**Comment by anisimovyuriy** [comment]
Dear tig2810 and IanYates,

First of all my apologies that I haven't provided explanations right away and put it as wonfix as I was disturbed. I did it on purpose as when you call Mapper.Compile() - which is the best practice use it when application starts - and only once - EVERYTHING will be compiled and stored in the cache and when you use actual Map function it's just a call of function and that is all. So once you call Mapper.Compile() - if there'd be any errors the exception will be thrown and you right away will know that something is bad with your registration.
Please let know if I got you in a correct way and thanks a lot guys for your input!

**Comment by IanYates** [comment]
@anisimovyuriy  - makes sense.  So we can just have a unit test set up all of our maps, then call Mapper.Compile() and pass if there's no exception.  Great :)

I've got a lot of hand-rolled mapping code and was looking at AutoMapper, ValueInjector, etc but your syntax seems quite nice and the website makes it easy to follow what's happening.  I'm looking forward to trying this out in the next couple of weeks!

Thanks :100:

**Comment by anisimovyuriy** [comment]
Thank you IanYates! And I hope you'll enjoy it and let us know if you have something in mind to improve Expressmapper - we'll appreciate it :+1:

**Comment by tig2810** [comment]
Hi there,
I'm not with either of you on this one.  Great, mapper.compile throws an exception, but the point i would have thought was to automatically highlight the issue without the need for separate tests.   I understand you dont want to do it, but i still think it would have been a great feature.

**Comment by anisimovyuriy** [comment]
Hey tig2810,

Thanks for your reply!
Let me take a look on that one more deeply. Maybe I've just misunderstood you and I do want implement it - of course it is reasonable. I'll let you know ASAP.

Thanks,
Yuriy

**Comment by tig2810** [comment]
Hi Yurly,

Many thanks for taking another look.   I only checked it for a few mins but for example, Mapper.Compile() will throw a stackoverflowexception but AssertConfigurationIsValid() will also show the mismapped fields.    It's a bit lazy of me perhaps but it's great when tools provide feedback and help you. 

I think given the great performance of Expressmapper, if it were to give some of the sugar i get with Automapper, I would probably just switch to using it exclusively. 

Thanks again

**Comment by DerAlbertCom** [comment]
Any Idea on a timeframe for 2.0? Configuration validation with detailed information about which mappings are not valid (as in AutoMapper) would be highly appreciated

**Comment by anisimovyuriy** [comment]
Dear DeAlbertCom,

There are some priorities like:
1. Hierarchy / Interface mapping support
2. Tracking object instances during mapping for recursive support

Afterwards Yes :) - so to answer your question it's approx. 2-3 months as I've been kept busy recently...

Thanks,
Yuriy

**Comment by mattzink** [comment]
Another vote for this feature. This would certainly allow us to consider switching from automapper.

**Comment by waynebrantley** [comment]
With AutoMapper AssertConfigurationIsValid() it checks to make sure that every single Destination type member has a corresponding type member on the source type.
With the Compile() on ExpressMapper it does not do anything that I can determine.

This code happily runs...

```
    public class SourceFull
    {
        public string StringValue1 { get; set; }
        public string StringValue2 { get; set; }
        public int IntValue1 { get; set; }
    }

    public class DestExact
    {
        public string StringValue1 { get; set; }
        public string StringValue2 { get; set; }
        public int IntValue1 { get; set; }
    }

    public class DestPartial
    {
        public string StringValue1 { get; set; }
        public int IntValue1 { get; set; }
    }

    public class DestMoreThanSource
    {
        public string StringValue1 { get; set; }
        public int IntValue1 { get; set; }
        public int IntValue2 { get; set; }
    }
```

```
    Mapper.Register<SourceFull, DestExact>();
    Mapper.Register<SourceFull, DestPartial>();
    Mapper.Register<SourceFull, DestMoreThanSource>();
    Mapper.Compile();
    var src = new SourceFull { IntValue1 = 12, StringValue1 = "Value1", StringValue2 = "Value2" };
    var deste = src.Map<SourceFull, DestExact>();
    var destp = src.Map<SourceFull, DestPartial>();
    var destm = src.Map<SourceFull, DestMoreThanSource>();
```

So, I am not sure what you guys mean when you say it will throw an exception.

What would be nice is something flexible like this:
```
    Mapper.Register<SourceFull, DestExact>().EnforceAllDestInSource();
    Mapper.Register<SourceFull, DestPartial>().EnforceAllSourceInDest();
```
Where the enforcers make sure there is an equivalent property to map.  If there are extra properties, if you specify them with `.Ignore()`  then it will pass the enforcement.

As a side note, I think this issue is the same request.  https://github.com/expressmapper/ExpressMapper/issues/33

**Comment by waynebrantley** [comment]
Quick and dirty extension to check mappings.
- Does not respect Ignore()   --not sure how to get this info from ExpressMapper
- Custom mapped properties (one name to another, custom calculations, etc) would not be respected.

Basically, just checks that the property names are all present - high level concept for what is needed.  
```
    public static class MapperExt
    {
        public static IMemberConfiguration<T, TN> AllDestPresent<T, TN>(this IMemberConfiguration<T, TN> config)
        {
            var sourceProps = typeof(T).GetProperties(BindingFlags.Instance | BindingFlags.Public);
            var destProps = typeof(TN).GetProperties(BindingFlags.Instance | BindingFlags.Public);

            var sourceNames = sourceProps.Select(x => x.Name).ToArray();
            var missingDestProps = destProps.Where(x => !sourceNames.Contains(x.Name)).Select(x => x.Name).ToList();
            if (missingDestProps.Count > 0)
                throw new Exception("Source is missing: " + string.Join(", ", missingDestProps.ToArray()));

            return config;
        }
        public static IMemberConfiguration<T, TN> AllSourcePresent<T, TN>(this IMemberConfiguration<T, TN> config)
        {
            var sourceProps = typeof(T).GetProperties(BindingFlags.Instance | BindingFlags.Public);
            var destProps = typeof(TN).GetProperties(BindingFlags.Instance | BindingFlags.Public);

            var destNames = destProps.Select(x => x.Name).ToArray();
            var missingSourceProps = sourceProps.Where(x => !destNames.Contains(x.Name)).Select(x => x.Name).ToList();
            if (missingSourceProps.Count > 0)
                throw new Exception("Source is missing: " + string.Join(", ", missingSourceProps.ToArray()));

            return config;
        }
    }
```
This will now throw an exception and list the property names that had issues.
```
            Mapper.Register<SourceFull, DestExact>().AllDestPresent();
            Mapper.Register<SourceFull, DestPartial>().AllDestPresent(); ;
            Mapper.Register<SourceFull, DestMoreThanSource>().AllDestPresent();
            Mapper.Compile();
```

I cannot imagine using a mapper without some sort of validation like this....unless I wrote unit tests for every mapping to do the validation.  However, sometimes mappings are only preformed in one place - and if I have to write a test and manually check the mappings, I might as well have hand written it to start with IMHO.

Thoughts?

**Comment by paranoidandroid522** [comment]
The main problem I currently have with ExpressMapper compared to AutoMapper is that `Compile() `doesn't check for unmapped properties in the destination type. This is very error prone -- AutoMapper's `AssertConfigurationIsValid()` avoids a very common bug that extremely easily slips by unnoticed (even with unit testing): Adding misspelled or otherwise non-mappable properties to the target type. 
AutoMapper helps you avoiding this bug by throwing an exception in `AssertConfigurationIsValid()` unless you explicitly `Ignore()` them or configure a custom mapping (except for getter-only properties); ExpressMapper's `Compile()` just silently ignores them. 

So my wishlist for ExpressMapper would be:

- `Compile()` should throw when it finds non-mappable properties in the target type, unless they are `Ignored()` or getter-only
- The exception thrown by `Compile()` should contain a clear, easily understandable message what is wrong with my mappings.

**Comment by karlcassar** [comment]
I was looking for the exact same functionality and came across this thread and would like to vote for this too.  I think it is very beneficial that a Mapper framework lets you know if you "forgot" to map a destination property, maybe it's an omission, spelling mistake, or something of that sort.
